torch/_ops.py

```diff
@@ -266,6 +266,28 @@ def __init__(self, name):
         for dispatch_key in _HIGHER_ORDER_OP_DEFAULT_FALLTHROUGH_DISPATCH_KEYS:
             self.fallthrough(dispatch_key)
 
+        # [NOTE] We have to register pre-dispatch key implementation
+        # because sometimes HOP use aot-dispatch tracing to detect certaion
+        # mutations. This is problematic when we are functionalizing HOP
+        # during pre-dispatch because when the inner tracer starts, it will see
+        # that PreDispatch key is still active. In that case, we just redispatch
+        # it to next key. This is only safe to do when PreDispatch key stack has no
+        # active modes.
+        def _(*args, **kwargs):
+            if _len_torch_dispatch_stack_pre_dispatch() == 0:
+                with torch._C._ExcludeDispatchKeyGuard(
+                    torch._C.DispatchKeySet(DispatchKey.PreDispatch)
+                ):
+                    return self(*args, **kwargs)
+            raise AssertionError(
+                """
+                Can't directly invoke HOP implementation at PreDispatch key
+                if there are active modes on PreDispatch mode stack.
+                """
+            )
+
+        self.py_impl(torch._C.DispatchKey.PreDispatch)(_)
```

Comment:
https://github.com/pytorch/pytorch/blob/b88621040a9fb6402ed02d2bb5e4ae6d2e4c1704/torch/_higher_order_ops/cond.py#L247 
Autograd seems like handled on op by op basis. But this implementation should exist for all predispatch keys

Reply:
Concretely this would look like having a fallbacks.py file, and some registration mechanism for fallbacks:

```
@HigherOrderOp.py_fallback(dk: DispatchKey_or_mode):
def pre_autograd_fallback(*args, **kwargs):
            if _len_torch_dispatch_stack_pre_dispatch() == 0:
                with torch._C._ExcludeDispatchKeyGuard(
                    torch._C.DispatchKeySet(DispatchKey.PreDispatch)
                ):
                    return self(*args, **kwargs)
            raise AssertionError(
                """
                Can't directly invoke HOP implementation at PreDispatch key
                if there are active modes on PreDispatch mode stack.
                """
            )
```
this registers `pre_autograd_fallback` to some global dict that we can then read from when dispatching on a HOP.